accord-core/src/main/java/accord/topology/TopologyManager.java

```diff
@@ -98,36 +97,33 @@ static class EpochState
 
             this.addedRanges = global.ranges.subtract(prevRanges).mergeTouching();
             this.removedRanges = prevRanges.mergeTouching().subtract(global.ranges);
-            this.prevSyncComplete = addedRanges.union(MERGE_ADJACENT, prevSyncComplete.subtract(removedRanges));
-            this.curSyncComplete = this.syncComplete = addedRanges;
+            this.syncComplete = addedRanges;
```

Comment:
because `prevSyncComplete` concept no longer exists.  

The logic is used when we launch transactions to see what older epochs we don't know the state of that intersect the transaction, that way we include them.  The target reason is token movements... we need to include the node that had the range removed until we hear from it that it knows the new epoch.

So, when a range is added, we know we don't need to track it as no older epochs have it as well
When a range is removed, it is no longer present... so we won't add to syncComplete...

does this make sense?

Reply:
Could you elaborate a bit on this? This is critical for protocol correctness so would be good to understand what is being removed. The comment appears to be against a dead link.

Reply:
`syncComplete` is based off `accord.local.Node#onTopologyUpdateInternal`  and `accord.api.ConfigurationService.EpochReady#coordination`

```
ready.coordination.addCallback(() ->  {
            synchronized (AbstractConfigurationService.this)
            {
                localSyncComplete(epochs.getOrCreate(ready.epoch).topology, startSync);
            }
        });
```

This logic was changed (we talked about this a long time ago) so that each epoch is now sequential and we don't allow epoch=N to happen-before epoch=N-1 (currently N can be syncComplete before N-1, simple case is that N-1 is a node joining and must stream, and N is a table comment update).

If you look at `accord.local.Node#onTopologyUpdateInternal` coordination is now sequenced so `N` happens-after `N-1`.

In Apache Cassandra this is notifies `localSyncComplete` which marks the epoch as `epochState.setSyncStatus(SyncStatus.NOTIFYING);` and notifies the *nodes in the topology* about the epoch.  Those nodes will do the same which cause `accord.topology.TopologyManager#onEpochSyncComplete` to get called, which updates `syncComplete` (and `prevSyncComplete` and `curSyncComplete`)...

Now we have to turn to 2 cases that exist in Apache Cassandra: Expand Cluster, and Host Replacement.  In both of these cases a *new node* starts to join and will know about past epochs (as it's needed for processing), but won't be present in those past epochs.  This leads us to a situation where those past epochs have `syncComplete = []` and we include every single epoch for coordination.

By sequencing `EpochReady.coordination`, once this new node *sees* epoch=N be sync complete by its peers, it *knows* epoch=N-1 is also syncComplete.

Without this change, we would require every node in the cluster to notify the joining node about every epoch they know and their sync status... if the sync is also not complete then the node would be stuck with a partial update as it wouldn't learn about new ones!